### PR TITLE
feat: context support, pointer of connection, exported ErrClosed, parameterized queries

### PR DESCRIPTION
**Pull request by aldy505**
Sorry for such a big git diff. But anyway, these are the changes that I made:
1. `gorqlite.Open()` now returns a pointer of `Connection`, this may break some users' experience. But it's only a pointer away. Every functionality are the same.
2. Addition of `QueryContext`, `QueryOneContext`, `QueueContext`, `QueueOneContext`, `WriteContext`, and `WriteOneContext` method.
  - Update 2022-09-18: Additional `QueryOneParameterizedContext`, `QueryParameterizedContext`, `WriteOneParameterizedContext`, `WriteParameterizedContext`, `QueueOneParameterizedContext`, and `QueueParameterizedContext` based on #26.
4. Comment-style from `/* */` to `//` to comply with how `go doc` generates its' documentation. So, every documented code will looks good on https://pkg.go.dev/github.com/rqlite/gorqlite
5. `errClosed = errors.New("gorqlite: connection is closed")` is now exported as `ErrClosed`, so users can do:
   ```go
    func main() {
      qr, err := conn.QueryContext(ctx, "some query...")
      if err != nil {
        if errors.Is(err, gorqlite.ErrClosed) {
          // do a retry because connection is closed
        }
      }
    }
   ```
6. (Removed, handled by #26) ~~Parameterized query is now supported, and it doesn't break the current API.~~
    <details>
    <summary>Old code spoiler</summary>

      ```go
      func main() {
        rows, err := conn.QueryContext(ctx, "SELECT name FROM table WHERE id = ?", 12)
        // do things...

        // and this also works
        secondRows, err := conn.QueryContext(ctx, "SELECT name FROM table WHERE id = 12")

        // also with Write() and Queue() methods (and their similar methods)
        _, err = conn.WriteContext(ctx, "INSERT INTO table (id, name) VALUES (?, ?)", 12, "James")

        _, err = conn.QueueContext(ctx, "INSERT INTO table (id, name) VALUES (?, ?)", 12, "James")
      }
    ```
    </details>
8. Support for scanning `bool` and `[]bytes` (handle `[]bytes` only, `bool` is handled by #26)
9. Exported consistency levels to constants `ConsistencyLevelNone`, `ConsistencyLevelWeak`, `ConsistencyLevelStrong`. Users don't have to open up documentation ever again for this one.
10. Rewritten tests to have better and consistent behavior

### CODE REVIEW

**Comment by aldy505** [comment]
The CI failed because `NewRequestWithContext` is available since go 1.13, yet the CI uses go 1.12 :/

Reference: https://pkg.go.dev/net/http#NewRequestWithContext

**Comment by otoolep** [comment]
Thanks for your contribution.

- We might need to version this, to prevent breaking changes.
- We could just bump the CircleCI version and all should work, right?

**Comment by otoolep** [comment]
Do you want to try modifying `.circleci/config.yml` so that it includes this line: https://github.com/rqlite/rqlite/blob/master/.circleci/config.yml#L9

That's the CircleCI image I used to test rqlite.

**Comment by otoolep** [comment]
That CircleCI image runs Go go1.16.7.

**Comment by aldy505** [comment]
@otoolep The CI works fine now. About the versioning, should we go directly to v2 or do you have something else in mind?

**Comment by otoolep** [comment]
v2 sounds reasonable -- can  you make that change?

**Comment by aldy505** [comment]
@otoolep I've updated the version. Everything should be ok, so far.

**Comment by otoolep** [comment]
Thanks for this -- I'm traveling until early August, so may be slow responding. Once I return to home base, I'll examine in detail.

**Comment by sgalsaleh** [comment]
@otoolep Any updates on this? It would be great to have this merged. We'd like to adopt it.

File: api.go
Comment (sgalsaleh) [inline api.go:144]: This loop could cause an "index out of range" error/panic. IMO, there should be some validation that the length of the `parameters` array matches the length of the `sqlStatements` array when parameters are passed.

File: query.go
Comment (sgalsaleh) [inline query.go:123]: This function expects an array of sqlStatements, but in your PR description, you have the following example:
```
rows, err := conn.QueryContext(ctx, "SELECT name FROM table WHERE id = ?", 12)
```
I guess I can't see how that would work since a `string` and not a `[]string` is being passed in the example. The same applies to other functions as well (e.g. `Query`, `Write`, and `WriteContext`). I also don't see that covered in the unit tests.

It could be the case that the PR description is not up-to-date, which is fine, but it would be great to have the official README updated with accurate usage instructions.

Comment (aldy505) [reply query.go:123]: Ah, thanks for the catch. I don't remember why I made it that way, but I'll take a look later this weekend.

**Comment by sgalsaleh** [comment]
Piggybacking on this PR since it doesn't seem that we can submit issues in this repo: would it be possible to support nullable types as well for the `Scan` function? I've tried using both `*string` and `sql.NullString` but neither of those worked. Happy to submit a PR for that after this PR is merged.

**Comment by aldy505** [comment]
With #26 merged, I guess I'm going to drop my parameterized query implementation, and use @sgalsaleh's instead.

**Comment by sgalsaleh** [comment]
Sounds good @aldy505. There are some changes in your PR that I did not include in mine that are probably still good to have:

> 2- Addition of QueryContext, QueryOneContext, QueueContext, QueueOneContext, WriteContext, and WriteOneContext method.

> 3- Comment-style from /* */ to // to comply with how go doc generates its' documentation. So, every documented code will looks good on https://pkg.go.dev/github.com/rqlite/gorqlite

> 8- Rewritten tests to have better and consistent behavior

It would be great to get those merged in. Would you be willing to rebase the PR, or would you rather create a separate PR and this one can be closed?

Note: I also added some tests in my PR and it'd be great to have them rewritten the same way.

**Comment by aldy505** [comment]
@sgalsaleh I'm actually in the middle of merge conflict between my branch and the upstream. I'll get this PR updated soon.

**Comment by sgalsaleh** [comment]
Ah, I see. Sounds good. Let me know when it's ready for review and I can review it and merge it.

**Comment by aldy505** [comment]
I think everything should be good now. If there are things to improve, it surely would be the tests, moreover, we should test how it handles scanning of different data types (example: string or text in rqlite to int16 in Go types).

At this point, I don't think we should go to the v2 versioning mark. Any thoughts? @otoolep

**Comment by otoolep** [comment]
If we don't need v2, sounds fine to me. 

@sgalsaleh -- review and merge away, thanks!

File: conn_test.go
Comment (sgalsaleh) [inline conn_test.go:9]: How about using `conn.ConsistencyLevel()` to validate that the consistency level was actually changed?

File: query.go
Comment (sgalsaleh) [inline query.go:567]: I don't believe `src` can ever be a bool, no? Since rqlite/sqlite does not have such a type. Maybe @otoolep can confirm.

File: query.go
Comment (sgalsaleh) [inline query.go:574]: I'd suggest using `[]byte` instead of `[]uint8` just to follow conventional Go code. `byte` is basically an alias for `uint8`

File: query_test.go
Comment (sgalsaleh) [inline query_test.go:52]: Tip: you can use `assert.ElementsMatch(t, expect, qr.Columns())` function from the `github.com/stretchr/testify/assert` package (not sure if that package exists in this project, but should be easy to add).

File: gorqlite_test.go
Comment (sgalsaleh) [inline gorqlite_test.go:36]: Since tests can fail and this won't necessarily run to clean up the database if that happens, there are two options:

1- Run this in a `defer` statement and place it before `m.Run()` so that it always gets executed.
2- Keep it here, but also call it at the beginning before creating the tables so that tests start with a fresh/clean database.

The only benefit of the second option is to give the user the ability to inspect the data in the database when tests fail, which I prefer.

File: write_test.go
Comment (sgalsaleh) [inline write_test.go:13]: This table isn't cleaned up afterwards.

Comment (sgalsaleh) [inline write_test.go:13]: This table isn't cleaned up afterwards.

File: write_test.go
Comment (sgalsaleh) [inline write_test.go:26]: How would this table exist if its name is random?

File: write_test.go
Comment (sgalsaleh) [inline write_test.go:63]: Ditto

File: write_test.go
Comment (sgalsaleh) [inline write_test.go:71]: Ditto

File: write_test.go
Comment (sgalsaleh) [inline write_test.go:223]: Let's use `t.Errorf` :)

File: write_test.go
Comment (sgalsaleh) [inline write_test.go:250]: Ditto

File: write_test.go
Comment (sgalsaleh) [inline write_test.go:254]: Ditto

File: gorqlite_test.go
Comment (sgalsaleh) [inline gorqlite_test.go:13]: I actually liked the fact that each test controlled its table schema and data. I feel like having a global connection/state will make it hard to track down what the current state of the database is when running tests. I would personally prefer the approach that was previously implemented where each test creates its own table and tears it down afterwards.

**Comment by sgalsaleh** [comment]
@aldy505 Thanks for rebasing and updating the PR. I left some comments, let me know if anything is not clear.

**Review by sgalsaleh — CHANGES_REQUESTED** [review]
@aldy505 Thanks for rebasing and updating the PR. I left some comments, let me know if anything is not clear.

File: gorqlite_test.go
Comment (sgalsaleh) [reply gorqlite_test.go:36]: Note: this won't be necessary if we revert to having each test drop/create its own table and tear it down afterwards.

File: query.go
Comment (sgalsaleh) [inline query.go:576]: I don't believe `src` can ever be `uint8`. As far as I know, it will either be a `int64`, `float64`, `string`, or `nil`. It might be good to handle all of those cases similar to other types.

File: query_test.go
Comment (aldy505) [reply query_test.go:52]: I don't want to use a third party package for just asserting that one.

File: query.go
Comment (aldy505) [reply query.go:574]: IIRC, I've tried `[]byte` before, I'm gonna re-check again if `[]byte` works.

File: gorqlite_test.go
Comment (aldy505) [reply gorqlite_test.go:36]: 1. If we use `os.Exit()`, any `defer` statement won't be run.
2. I'll do the second option.

File: write_test.go
Comment (aldy505) [reply write_test.go:223]: I thought I was gonna leave this one out for you, haha! Okay, I'll tidy this one out 😄

File: gorqlite_test.go
Comment (aldy505) [reply gorqlite_test.go:13]: The problem with that is it's hard to test with `-race` flag. The current CircleCI test uses `go test -timeout 60s -v ./...`. Yet, I often uses `go test -timeout 60s -v -cover -covermode=atomic -race ./...`. If we constantly change schemas for the same table name, creating and dropping again, it doesn't make sense for that kind of test.

We might still have one connection tho, but the tables are created and dropped for each test suite, how about that?

File: query.go
Comment (sgalsaleh) [reply query.go:574]: Ah I see. That's fine then, let's keep `uint8`.

Comment (sgalsaleh) [reply query.go:574]: Have you tried defining the destination parameter to be a `[]byte` instead of `[]uint8` when calling the `Scan` function? With that being said, I don't mind keeping `[]uint8`.

File: gorqlite_test.go
Comment (sgalsaleh) [reply gorqlite_test.go:13]: > We might still have one connection tho, but the tables are created and dropped for each test suite, how about that?

Yep, that sounds great.

File: write_test.go
Comment (sgalsaleh) [reply write_test.go:223]: Haha. Thanks!

File: gorqlite_test.go
Comment (sgalsaleh) [reply gorqlite_test.go:36]: I guess this is no longer relevant since creating/dropping tables will be moved to the test suites.

File: query_test.go
Comment (sgalsaleh) [reply query_test.go:52]: Ah okay.

File: query.go
Comment (otoolep) [reply query.go:567]: As per https://www.sqlite.org/datatype3.html:

> 2.1. Boolean Datatype
> 
> SQLite does not have a separate Boolean storage class. Instead, Boolean values are stored as integers 0 (false) and 1 (true).
> 
> SQLite recognizes the keywords "TRUE" and "FALSE", as of version 3.23.0 (2018-04-02) but those keywords are really just alternative spellings for the integer literals 1 and 0 respectively. 

Therefore rqlite does not have a boolean type.

**Comment by aldy505** [comment]
Whoa, something's wrong with Github's review requests.

Anyway, sorry for a bit of a hiatus, it's been a busy week on my office and on my other side projects. These changes should be aligned with what's reviewed before.

File: README.md
Comment (sgalsaleh) [inline README.md:31]: this is no longer accurate, correct?

File: api.go
Comment (sgalsaleh) [inline api.go:71]: why is this change needed?

File: query.go
Comment (sgalsaleh) [inline query.go:583]:
```suggestion
				*d = []byte(src)
```

File: query_test.go
Comment (sgalsaleh) [inline query_test.go:24]: I thought we were gonna keep creating/dropping tables per test

File: api.go
Comment (aldy505) [reply api.go:71]: `ioutil` is deprecated per 1.16 https://pkg.go.dev/io/ioutil#ReadAll

File: README.md
Comment (aldy505) [reply README.md:31]: ah right, I forgot

File: query_test.go
Comment (aldy505) [reply query_test.go:24]: Yes, we are. See line 38-48 of this file

Comment (sgalsaleh) [reply query_test.go:24]: Ah I missed that, thanks.

File: api.go
Comment (sgalsaleh) [reply api.go:71]: Ah, good to know!

**Review by sgalsaleh — APPROVED** [review]
LGTM

**Comment by maximenrb** [comment]
Hi, any news about this RP? It would be very great to have context support

**Review by sgalsaleh — APPROVED** [review]
LGTM after resolving conflicts

**Comment by sgalsaleh** [comment]
@maxnrb merged

**Comment by otoolep** [comment]
Are we sure about this? Tests seem broken now, though that might be due to the CircleCI security issue that meant all keys were rotated. 

Do we know that this PR passes all tests? @sgalsaleh

**Comment by sgalsaleh** [comment]
Hmm, tests passed on the PR, I'll take a look shortly.

**Comment by sgalsaleh** [comment]
The tests passed on the last commit in the PR before I merged: https://app.circleci.com/pipelines/github/rqlite/gorqlite/81/workflows/58f15b14-af6c-4939-959b-cca322d25383/jobs/84

**Comment by otoolep** [comment]
OK, let me look into the CircleCI config.

**Comment by aldy505** [comment]
Hi, sorry I've been busy with work today. Is there any other problem regarding tests?

**Comment by otoolep** [comment]
No, all good once I fixed up CircleCI.

**Pull request merged**
